### DISCUSSION TITLE
Does MarkdownUI support base64 images embedded in the markdown text?

### DISCUSSION THREAD

**Discussion by michzio** [gonzalezreal/swift-markdown-ui#205] [Q&A]
I have such markdown like: 

```
test priorytet średni wiadomość   ![img](data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACkAAAAoCAMAAABU4iNhAAABI1BMVEUAAAAAVf8AQL8AVdUASdsAQN8ARtEVQNUSSdsNTdkMSdsMRtwISt4ISNcIRtgIS9oOR9wOTN0NSt0NSNgNRtkMS9oMR9sMS9wLSt0LSN4LR9kLR9oMSdsMSNsMStsMSNwLStoLSNoLStoKSNsKStsMSdsMSNsMStsLSNsLStsLSdwLSdwLSNoLSdoLSdsKSdwKSNoKSdsMStsLSNsLSdsLSdwLStwLSdsMStwMSdoMSdsLStsLSdsLStwKStwMSdoMSNsLSdsLSNsLSdsKSdsMSdsMSdsLSdsLSdsLSdsLSdsLSdsLSNoLSdoLSdsLSdsLSdsLSdsLStsLSdwLSdsLSdsLSdwLSdsLSdsLSdsKSdsLSdsLSdsLSdsLSdsLSdv///9551VoAAAAX3RSTlMAAwQGBwgLDA4UFRYfICEiJCUmJygpKywtLi9EVFVWWFpgYWNkaWprcXJzdHV2d3t8foCGiImKjJiZmpyen62usLK+v8TGx8vMzc7P0NHT1NXW3t/g5ebp6uvz9Pv8/ul5oqcAAAABYktHRGDFt3wQAAABK0lEQVQ4y+3Vx1ICQRhF4YuiKEbAAJgwoWIEE5gFVFRMJDPn/d/CBYgWNLQLd3pWU11f1V8z010tSZJ80fPSO429l86iPn3Vc2hSNX3g/oSBApQv4lum4pky5P0VOPAMR141y3cMT32S1HZDeVGtisCVQ9IcbKp1OzAj6YFcu0U689xJPojI1gp4FIIRq/TDlJah0ypdEFEM6tcXXsP1SxAzySTJH8oUqT8ng2OVsmSrT0GznDTs9wmjnDDIcfP0QMP0wP8/+oWdHH6ZN8kIuKwnrguWFILRn51iLyxb5SoMSvfknRbYUeBW0ixsW+QuTEtyXNvmr0LGIUn9T3Ay1NQNn8Jjb/XN8kAmEVtrLJa4BHK1r+Peb3UjvO11fxvh2UgXjbdMMb3uqZAP6qWffXq1K0wAAAAASUVORK5CYII=)
```

The Markdown view doesn't render this image at all. 

[UPDATE]

Ok MarkdownUI doesn't support base64 images by default but it can be achived by writing custom ImageProvider in MarkdownUI 2.0 
like Base64ImageProivder and using some library like SVGKit. 

But I've encounter another problem, i.e. MarkdownUI detects only images defined in place like in above markdown 
but it doesn't parse images defined via reference like below 

```
Some text goes here.  [Link](https://google.com)  ![image][reference]   [reference]:data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDQwIiBoZ(...)WlnaH=
```

Is it possible to fix it this way the above image will be detected and base64 data uri returned to ImageProvider makeImage() function?

**Comment by michzio** [comment]
Ok I've found there is also InlineImageProvider and when I implemented it do parse base64 using SVGkit it started to work.

**Comment by gonzalezreal** [comment]
Hi @michzio,

MarkdownUI is correctly parsing image references. This code works as expected:

```swift
Markdown {
    """
    ![Puppy][ref]

    [ref]: https://picsum.photos/id/237/200/300
    """
}
```

I will see how I can add support for base64 images.

Thanks for your feedback!

**Comment by daver234** [comment]
Hi, can you tell me more about how you implemented Base64ImageProivder using SVGKit.
